### PR TITLE
Add unit test for generator support

### PR DESCRIPTION
**Pull request by poteat**
## Add unit test for generator support

I added, under `./test/serialization/`, a harness to test for serialization support of generators, including in-progress generators.

For brief background, generators are a ES6 and up JS/TS language feature (which polyfills to ES5 and below) utilizing the `yield` keyword:

```ts
function* myGenerator() {
  yield "I am running";
  yield "I am running again"
}

const gen = myGenerator();

gen.next(); // I am running
gen.next(); // I am running again
```

By in-progress generator, I mean an actual Generator object (as opposed to the constructor function) that has executed at least one step.

### CODE REVIEW

File: ts-closure-transform/test/serialization/generator.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  return deserialize(serialize(value));
+}
+
+export function* myGenerator() {
+  yield 42;
+  yield 84;
+}
+
+export function simpleGenerator() {
+  expect(myGenerator().next()).to.equal(42);
```
Comment (lhoste-bell) [inline ts-closure-transform/test/serialization/generator.ts:14]: This should be `expect(myGenerator().next()).to.deep.equal({ value: 42, done: false });`

File: ts-closure-transform/test/serialization/generator.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+export function roundtripGeneratorConstructor() {
+  expect(roundtrip(myGenerator)().next()).to.equal(42);
```
Comment (lhoste-bell) [inline ts-closure-transform/test/serialization/generator.ts:18]: ^

File: ts-closure-transform/test/serialization/generator.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+export function roundtripGeneratorInProgress() {
+  const gen = myGenerator();
+  expect(gen.next()).to.equal(42);
```
Comment (lhoste-bell) [inline ts-closure-transform/test/serialization/generator.ts:23]:
```
    let gen = myGenerator();
    expect(gen.next()).to.deep.equal(42);
    gen = roundtrip(gen);
    expect(gen.next()).to.deep.equal({ value: 84, done: false });
    gen = roundtrip(gen);
    expect(gen.next()).to.deep.equal({ value: undefined, done: true });

File: ts-closure-transform/test/serialization/generator.ts (hunk 1 above)
Comment (poteat) [reply ts-closure-transform/test/serialization/generator.ts:14]: Good catch, agree.

**Comment by lhoste-bell** [comment]
Hi @poteat,

Recent patches on master enable the unit test in https://github.com/nokia/ts-serialize-closures/commit/a92161dbb2044b08c5fe94cf8382b6734099f9e3 to pass. However I still need to investigate why our transformer does not visit the emitted generator from tsc, hence the manual inclusion of the `__generator` function. The same issue applies to the serialization of awaited async functions.

**Comment by poteat** [comment]
Sure thing!  I have since realized there might be another weird AST case for combinations of async/await and yield, i.e. `AsyncGenerator`.

Obviously we can close this PR if you've already applied the fixes to these tests on a separate branch.

**Pull request closed without merging**